store/v2/root/store_test.go

```diff
@@ -0,0 +1,958 @@
+		[]string{"0 2", "0 1", "0 0", "0"})
+	testCase(t, store.ReverseIterator([]byte{0}, []byte{0}), []string{}) // start = end
+	testCase(t, store.ReverseIterator([]byte{1}, []byte{0}), []string{}) // start > end
+
+	testCase(t, types.KVStorePrefixIterator(store, []byte{0}),
+		[]string{"0", "0 0", "0 1", "0 2"})
+	testCase(t, types.KVStoreReversePrefixIterator(store, []byte{0}),
+		[]string{"0 2", "0 1", "0 0", "0"})
+
+	require.Panics(t, func() { store.Iterator([]byte{}, nil) }, "Iterator(empty key) should panic")
+	require.Panics(t, func() { store.Iterator(nil, []byte{}) }, "Iterator(empty key) should panic")
+	require.Panics(t, func() { store.ReverseIterator([]byte{}, nil) }, "Iterator(empty key) should panic")
+	require.Panics(t, func() { store.ReverseIterator(nil, []byte{}) }, "Iterator(empty key) should panic")
+}
+
+func TestCommit(t *testing.T) {
+	testBasic := func(opts StoreConfig) {
+		// Sanity test for Merkle hashing
+		db := memdb.NewDB()
+		store, err := NewStore(db, opts)
+		require.NoError(t, err)
+		require.Zero(t, store.LastCommitID())
+		idNew := store.Commit()
+
+		// Adding one record changes the hash
+		s1 := store.GetKVStore(skey_1)
+		s1.Set([]byte{0}, []byte{0})
+		idOne := store.Commit()
+		require.Equal(t, idNew.Version+1, idOne.Version)
+		require.NotEqual(t, idNew.Hash, idOne.Hash)
+
+		// Hash of emptied store is same as new store
+		s1.Delete([]byte{0})
+		idEmptied := store.Commit()
+		require.Equal(t, idNew.Hash, idEmptied.Hash)
+
+		previd := idOne
+		for i := byte(1); i < 5; i++ {
+			s1.Set([]byte{i}, []byte{i})
+			id := store.Commit()
+			lastid := store.LastCommitID()
+			require.Equal(t, id.Hash, lastid.Hash)
+			require.Equal(t, id.Version, lastid.Version)
+			require.NotEqual(t, previd.Hash, id.Hash)
+			require.NotEqual(t, previd.Version, id.Version)
+		}
+	}
+	basicOpts := simpleStoreConfig(t)
+	basicOpts.Pruning = types.PruneNothing
+	testBasic(basicOpts)
+	basicOpts.StateCommitmentDB = memdb.NewDB()
+	testBasic(basicOpts)
+
+	testFailedCommit := func(t *testing.T,
+		store *Store,
+		db dbm.DBConnection,
+		opts StoreConfig) {
+		if db == nil {
+			db = store.stateDB
+		}
+		s1 := store.GetKVStore(skey_1)
+		s1.Set([]byte{0}, []byte{0})
+		require.Panics(t, func() { store.Commit() })
+		require.NoError(t, store.Close())
+
+		// No version should be saved in the backing DB(s)
+		versions, _ := db.Versions()
+		require.Equal(t, 0, versions.Count())
+		if store.StateCommitmentDB != nil {
+			versions, _ = store.StateCommitmentDB.Versions()
+			require.Equal(t, 0, versions.Count())
+		}
+
+		// The store should now be reloaded successfully
+		store, err := NewStore(db, opts)
+		require.NoError(t, err)
+		s1 = store.GetKVStore(skey_1)
+		require.Nil(t, s1.Get([]byte{0}))
+		require.NoError(t, store.Close())
+	}
+
+	opts := simpleStoreConfig(t)
+	opts.Pruning = types.PruneNothing
+
+	// Ensure Store's commit is rolled back in each failure case...
+	t.Run("recover after failed Commit", func(t *testing.T) {
```

Comment:
I would like more test breakdowns like this.  These r.Runs are readable and descriptive.

Reply:
Ok I can look at adding more